File: cpp/src/local_planning/simulation.cpp
#include "navigation/local_planning/simulation.hpp"

#include <algorithm>
#include <chrono>
#include <cmath>
#include <map>

#include "navigation/local_planning/geometry.hpp"

namespace navigation::local_planning {

core::Pose integrate_unicycle(const core::Pose& pose, const core::VelocityCommand& cmd,
                              double dt) {
  // Below this |omega| the arc radius v/omega blows up numerically before the
  // path shape actually straightens out; a straight-line step at 1e-9 rad/s
  // is indistinguishable from the arc solution to any solver's precision.
  if (std::abs(cmd.omega) < 1e-9) {
    return core::Pose{pose.x + cmd.v * dt * std::cos(pose.theta),
                      pose.y + cmd.v * dt * std::sin(pose.theta), pose.theta};
  }
  double theta_next = pose.theta + cmd.omega * dt;
  double x = pose.x + (cmd.v / cmd.omega) * (std::sin(theta_next) - std::sin(pose.theta));
  double y = pose.y - (cmd.v / cmd.omega) * (std::cos(theta_next) - std::cos(pose.theta));
  return core::Pose{x, y, wrap_to_pi(theta_next)};
}

SimResult simulate(core::LocalPlanner<core::ObstacleQuery>& planner, core::ObstacleQuery& space,
                   const core::RobotState& start, const core::LocalTask& task,
                   const SimConfig& config, core::TraceRecorder* recorder) {
  auto t_start = std::chrono::steady_clock::now();

  planner.reset();
  const core::Footprint footprint{config.footprint_radius};

  SimResult result;
  result.trajectory.push_back(start.pose);
  double min_clearance = space.distance_to_nearest(core::Point{start.pose.x, start.pose.y});
  double distance_traveled = 0.0;

  SimStatus status = SimStatus::TIMEOUT;
  int steps = 0;

  if (space.is_collision(footprint, start.pose)) {
    status = SimStatus::COLLISION;
  } else {
    core::RobotState state = start;
    for (steps = 1; steps <= config.max_steps; ++steps) {
      core::VelocityCommand cmd =
          planner.compute_command(space, state, task, config.control_dt, recorder);
      core::Pose pose = integrate_unicycle(state.pose, cmd, config.control_dt);

      if (recorder) {
        recorder->robot_moved(
            pose, core::TraceRecorder::EventData{{"v", cmd.v}, {"omega", cmd.omega}});
      }

      if (space.is_collision(footprint, pose)) {
        status = SimStatus::COLLISION;
        break;
      }

      min_clearance =
          std::min(min_clearance, space.distance_to_nearest(core::Point{pose.x, pose.y}));
      distance_traveled += std::hypot(pose.x - state.pose.x, pose.y - state.pose.y);
      result.trajectory.push_back(pose);

      if (std::hypot(pose.x - task.goal.x, pose.y - task.goal.y) <= config.goal_tolerance) {
        status = SimStatus::REACHED;
        break;
      }

      if (steps >= config.stall_window) {
        const core::Pose& past =
            result.trajectory[static_cast<size_t>(steps - config.stall_window)];
        if (std::hypot(pose.x - past.x, pose.y - past.y) < config.stall_distance) {
          status = SimStatus::STALLED;
          break;
        }
      }

      state = core::RobotState{pose, cmd.v, cmd.omega};
    }
    if (steps > config.max_steps) steps = config.max_steps;  // loop ran out -> TIMEOUT
  }

  result.status = status;
  result.success = (status == SimStatus::REACHED);
  result.steps = steps;
  result.time_to_goal = steps * config.control_dt;
  result.distance_traveled = distance_traveled;
  result.min_clearance = min_clearance;

  if (recorder) {
    if (result.success) recorder->path_found(result.trajectory);
    double runtime_sec =
        std::chrono::duration<double>(std::chrono::steady_clock::now() - t_start).count();
    std::map<std::string, double> metrics{
        {"time_to_goal", result.time_to_goal},
        {"distance_traveled", result.distance_traveled},
        {"min_clearance", result.min_clearance},
        {"steps", static_cast<double>(result.steps)},
        {"runtime_sec", runtime_sec},
        {"collided", status == SimStatus::COLLISION ? 1.0 : 0.0},
        {"stalled", status == SimStatus::STALLED ? 1.0 : 0.0},
    };
    recorder->planning_finished(result.success, metrics);
  }
  return result;
}

}  // namespace navigation::local_planning

File: python/navigation/local_planning/simulation.py
"""Closed-loop simulator: the execution harness every local planner runs under.

Depends on `core` abstractions only (LocalPlanner, ObstacleQuery, types, trace) --
no algorithm module import, no concrete map class. Shared by demos, tests, and
bench so the tick order (and therefore trace event stream) is identical no
matter which caller drives a planner.
"""

from __future__ import annotations

import enum
import math
import time
from dataclasses import dataclass, field

from navigation.core.capabilities import ObstacleQuery
from navigation.core.planner import LocalPlanner
from navigation.core.trace import TraceRecorder
from navigation.core.types import Footprint, LocalTask, Pose, RobotState, VelocityCommand

from ._geometry import wrap_to_pi

# Below this angular rate the closed-form arc (division by omega) is numerically
# unstable, so integration falls back to the straight-line limit instead --
# not a tolerance the caller can tune, so it is not a config parameter.
_OMEGA_EPS = 1e-9


class SimStatus(enum.Enum):
    REACHED = "reached"
    COLLISION = "collision"
    # Net displacement over a sliding window fell below tolerance: a distinct
    # failure mode from TIMEOUT (e.g. Potential Fields local-minimum lock-up),
    # worth reporting separately rather than lumping into "ran out of steps".
    STALLED = "stalled"
    TIMEOUT = "timeout"


@dataclass(frozen=True)
class SimConfig:
    control_dt: float
    max_steps: int
    goal_tolerance: float
    footprint_radius: float
    stall_window: int
    stall_distance: float


@dataclass
class SimResult:
    status: SimStatus
    success: bool
    steps: int
    time_to_goal: float  # steps * control_dt
    distance_traveled: float
    min_clearance: float
    trajectory: list[Pose] = field(default_factory=list)  # includes the start pose


def integrate_unicycle(pose: Pose, cmd: VelocityCommand, dt: float) -> Pose:
    """Exact constant-(v, omega) unicycle arc over one tick (closed form).

    A constant-velocity-and-turn-rate segment has a closed-form circular-arc
    solution, so this is exact for any dt -- Euler discretization error never
    mixes into the tick loop, keeping observed behavior (PF oscillation, PP
    tracking error) attributable to the algorithm rather than the integrator.
    """
    x, y, theta = pose
    v, omega = cmd.v, cmd.omega
    if abs(omega) < _OMEGA_EPS:
        return (x + v * dt * math.cos(theta), y + v * dt * math.sin(theta), theta)
    new_theta = theta + omega * dt
    x2 = x + (v / omega) * (math.sin(new_theta) - math.sin(theta))
    y2 = y - (v / omega) * (math.cos(new_theta) - math.cos(theta))
    return (x2, y2, wrap_to_pi(new_theta))


def _xy_dist(a: Pose, b: tuple[float, float]) -> float:
    return math.hypot(a[0] - b[0], a[1] - b[1])


def _finish(
    recorder: TraceRecorder | None,
    status: SimStatus,
    steps: int,
    control_dt: float,
    distance_traveled: float,
    min_clearance: float,
    trajectory: list[Pose],
    t_start: float,
) -> SimResult:
    success = status is SimStatus.REACHED
    time_to_goal = steps * control_dt
    metrics = {
        "time_to_goal": time_to_goal,
        "distance_traveled": distance_traveled,
        "min_clearance": min_clearance,
        "steps": float(steps),
        "runtime_sec": time.monotonic() - t_start,
        # additionalProperties on the trace `metrics` field is numeric-only, so
        # the outcome is carried as 0/1 flags instead of the SimStatus string.
        "collided": 1.0 if status is SimStatus.COLLISION else 0.0,
        "stalled": 1.0 if status is SimStatus.STALLED else 0.0,
    }
    if recorder is not None:
        if success:
            recorder.path_found(trajectory)
        recorder.planning_finished(success, metrics)
    return SimResult(
        status=status,
        success=success,
        steps=steps,
        time_to_goal=time_to_goal,
        distance_traveled=distance_traveled,
        min_clearance=min_clearance,
        trajectory=trajectory,
    )


def simulate(
    planner: LocalPlanner[ObstacleQuery],
    space: ObstacleQuery,
    start: RobotState,
    task: LocalTask,
    config: SimConfig,
    recorder: TraceRecorder | None = None,
) -> SimResult:
    """Run `planner` in closed loop against `space` from `start` until the
    episode ends (goal reached, collision, stall, or step budget exhausted)."""
    t_start = time.monotonic()
    planner.reset()
    footprint = Footprint(config.footprint_radius)
    min_clearance = space.distance_to_nearest((start.pose[0], start.pose[1]))
    trajectory: list[Pose] = [start.pose]
    if space.is_collision(footprint, start.pose):
        return _finish(
            recorder,
            SimStatus.COLLISION,
            0,
            config.control_dt,
            0.0,
            min_clearance,
            trajectory,
            t_start,
        )

    state = start
    distance_traveled = 0.0
    status = SimStatus.TIMEOUT
    steps = config.max_steps
    for step in range(1, config.max_steps + 1):
        cmd = planner.compute_command(space, state, task, config.control_dt, recorder)
        new_pose = integrate_unicycle(state.pose, cmd, config.control_dt)
        if recorder is not None:
            recorder.robot_moved(new_pose, data={"v": cmd.v, "omega": cmd.omega})
        if space.is_collision(footprint, new_pose):
            status = SimStatus.COLLISION
            steps = step
            break
        min_clearance = min(min_clearance, space.distance_to_nearest((new_pose[0], new_pose[1])))
        distance_traveled += _xy_dist(new_pose, (state.pose[0], state.pose[1]))
        trajectory.append(new_pose)
        if _xy_dist(new_pose, (task.goal[0], task.goal[1])) <= config.goal_tolerance:
            status = SimStatus.REACHED
            steps = step
            break
        if step >= config.stall_window:
            ref = trajectory[step - config.stall_window]
            if _xy_dist(new_pose, (ref[0], ref[1])) < config.stall_distance:
                status = SimStatus.STALLED
                steps = step
                break
        state = RobotState(new_pose, cmd.v, cmd.omega)

    return _finish(
        recorder,
        status,
        steps,
        config.control_dt,
        distance_traveled,
        min_clearance,
        trajectory,
        t_start,
    )
